src/Microsoft.Azure.WebJobs.Host/Timers/IWebJobsExceptionHandler.cs

```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Runtime.ExceptionServices;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.WebJobs.Host.Timers
+{
+    /// <summary>
+    /// Represents an unhandled exception handler.
+    /// </summary>
+    public interface IWebJobsExceptionHandler
+    {
+        /// <summary>
+        /// Called during host creation. Allows a reference to the host.
+        /// </summary>
+        /// <param name="host">The JobHost.</param>
+        void Initialize(JobHost host);
+
+        /// <summary>
+        /// Called when a timeout occurs.
+        /// </summary>
+        /// <returns>A Task.</returns>
+        Task OnTimeoutExceptionAsync(ExceptionDispatchInfo exceptionInfo, TimeSpan timeoutGracePeriod);
+
+        /// <summary>
+        /// Implement this method to handle an unhandled exception.
+        /// </summary>
+        /// <param name="exceptionInfo">The <see cref="ExceptionDispatchInfo"/>.</param>
+        Task OnUnhandledExceptionAsync(ExceptionDispatchInfo exceptionInfo);
+    }
```

Comment:
I want to call out this change.

I realized that simply re-throwing a FunctionTimeoutException wasn't very useful if we weren't able to cancel the shutdown token as well -- that will alert all *other* running functions that they should clean up because we're about to shut down.

So I decided to change this interface to look like this -- it allows you to grab the host during initialization, and then call StopAsync/Dispose on it to inform everyone else they should shut down. This would also stop all Listeners from trying to pick up anything new during the grace period.

You can see the default implementation in WebJobsExceptionHandler -- I'm writing tests for that now but wanted some feedback ASAP.

It feels wrong to have something that the JobHost created be responsible for killing it, but it seems to work. I'm open to other suggestions, too.